File: python/34_在排序数组中查找元素的第一个和最后一个位置.py
class Solution_34:
    def searchRange(self,nums,target):
        n=len(nums)
        l,r=0,n-1
        pos=-1
        ret=[]
        while l<=r:
            mid=(l+r)//2
            if(nums[mid]>target):
                r=mid-1
            elif(nums[mid]<target):
                l=mid+1
            else:
                pos=mid
                break
        ll, rr = pos, pos
        if(pos!=-1):
            while(ll-1>=0 and nums[ll-1]==nums[ll]):
                ll=ll-1
            while(rr+1<n and nums[rr+1]==nums[rr]):
                rr=rr+1
            ret.append(ll)
            ret.append(rr)
            return ret
        else:
            return [-1,-1]
if __name__ == '__main__':
    nums=eval(input("请输入一个数组"))
    target=eval(input("请输入目标值"))
    s_64=Solution_34()
    #list=s_64.searchRange(nums,target)
    #print(list)
    li=s_64.searchRange(nums,target)
    print(li)
    #print(s_64.searchRange(nums,target))
    print('\n')

File: c++/34_在排序数组中查找元素的第一个和最后一个位置.cpp
#include<iostream>
#include<vector>
using namespace std;
class Solution_34 {
	public:
	vector<int> searchRange(vector<int>& nums, int target) {
		int n = nums.size();
		int l = 0, r = n - 1;
		int k=-1;
		while (l <= r) {
			int mid = (l + r) >> 1;
			if (nums[mid] == target) {
				k = mid; break;
			}
			if (nums[mid] > target) {
				r = mid - 1;
			}
			else
			{
				l = mid - 1;
			}
		}
		if (k != -1) {
			int ll=k, rr=k;
			while (ll - 1 >= 0 && nums[ll - 1] == nums[ll]) {
				--ll;
			}
			while (rr + 1 < n && nums[rr] == nums[rr + 1]) {
				++rr;
			}
			return { ll,rr };
		}
		return{ -1,-1 };
	}
};
void main_34() {
	int num;
	vector<int> nums;
	int target;
	cout << "请输入一个数组:" << endl;
	while (cin >> num) {
		nums.push_back(num);
		if (cin.get() == '\n') break;
	}
	cout << "请输入要查找的数：" << endl;
	cin >> target;
	Solution_34 s34;
	vector<int> ret(s34.searchRange(nums, target));
	cout << "查找结果区间为：" << endl;
	cout << "[" << ret[0] << "," << ret[1] << "]" << endl;
	//ret=s34.searchRange(nums, target);
}